packages/bot/src/scratch/hooks/toolbox.js

```diff
@@ -80,122 +80,135 @@ Blockly.Toolbox.prototype.populate_ = function (newTree) {
-        Object.keys(blocks).forEach(blockKey => {
-            let keywords = ` ${blockKey}`;
-            const block = blocks[blockKey];
+                const only_block_contents = contents.filter(content => content.tagName.toUpperCase() === 'BLOCK');
+                return only_block_contents;
+            })
+            .flat();
+
+        block_contents.forEach(block_content => {
+            const block_type = block_content.getAttribute('type');
+            const block = Blockly.Blocks[block_type];
             const block_meta = block.meta instanceof Function && block.meta();
-            const block_definition = block.definition instanceof Function && block.definition();
+            const block_definitions = block.definition instanceof Function && block.definition();
 
             if (!block_meta) {
                 return;
             }
+            
+            // block_name matched
+            if (block_type.toUpperCase().includes(search_term)) {
+                flyout_content.unshift(block_content);
+                return;
+            }
 
-            Object.keys(block_meta).forEach(key => {
-                const meta = block_meta[key];
-                keywords += ` ${meta}`;
-            });
+            // block_meta matched
+            const matched_meta = Object.keys(block_meta)
+                .find(key => block_meta[key].toUpperCase().includes(search_term));
+            if (matched_meta && matched_meta.length) {
+                flyout_content.push(block_content);
+                return;
+            }
 
-            Object.keys(block_definition).forEach(key => {
-                const definition = block_definition[key];
+            // block_definition matched
+            const definition_key_to_search = /^((message)|(tooltip)|(field_dropdown$))/;
+            // eslint-disable-next-line consistent-return
+            const matched_definition = Object.keys(block_definitions).filter(key => {
+                const definition = block_definitions[key];
 
-                if (typeof definition === 'string') {
-                    keywords += ` ${definition}`;
-                } else if (definition instanceof Array) {
+                if (definition_key_to_search.test(key) &&
+                    definition.toUpperCase().includes(search_term)) {
+                    return true;
+                }
+
+                if (definition instanceof Array) {
+                    let has_dropdown_and_in_search = false;
+                    // eslint-disable-next-line consistent-return
                     definition.forEach(def => {
-                        if (def instanceof Object) {
-                            keywords += !def.type.includes('image') ? ` ${JSON.stringify(def)}` : '';
-                        } else {
-                            keywords += ` ${def}`;
+                        if (def.type === 'field_dropdown' &&
+                            JSON.stringify(def).toUpperCase().includes(search_term)) {
```

Comment:
We can put the stringified version of this property in a variable first for clarity, e.g.

```JavaScript
const stringified_definition = JSON.stringify(def).toUpperCase();
if (def.type === 'field_dropdown' && stringified_definition.includes(search_term)) {
    // Logic
}
```